### PR TITLE
[oneDPL] Transform_if APIs

### PR DESCRIPTION
**Pull request by danhoeflinger**
Proposing `oneapi::dpl::transform_if` specification for oneDPL.

### CODE REVIEW

File: source/elements/oneDPL/source/parallel_api.rst
Comment (akukanov) [inline source/elements/oneDPL/source/parallel_api.rst:699]: Let's use `typename` for template type parameters, for consistency with other APIs on this page.

File: source/elements/oneDPL/source/parallel_api.rst
Comment (akukanov) [inline source/elements/oneDPL/source/parallel_api.rst:694]: For a few other APIs here, the returned value is described - perhaps makes sense to do for `transform_if` as well.

File: source/elements/oneDPL/source/parallel_api.rst
Comment (danhoeflinger) [reply source/elements/oneDPL/source/parallel_api.rst]: applied suggestion

File: source/elements/oneDPL/source/parallel_api.rst
Comment (danhoeflinger) [reply source/elements/oneDPL/source/parallel_api.rst]: done.

File: source/elements/oneDPL/source/parallel_api.rst
Comment (danhoeflinger) [reply source/elements/oneDPL/source/parallel_api.rst:695]: I've added a sentence about the return for each overload, thanks.

File: source/elements/oneDPL/source/parallel_api.rst
Comment (akukanov) [inline source/elements/oneDPL/source/parallel_api.rst:715]:
```suggestion
initial value. The return value is an iterator targeting past the last considered element
of the output sequence, that is, :code:`result + (start - end)`.
```
Alternatively, we can use `std::advance(result, std::distance(start, end))` - but all sequences are random access anyway, so that seems excessive.

File: source/elements/oneDPL/source/parallel_api.rst
Comment (akukanov) [inline source/elements/oneDPL/source/parallel_api.rst:724]: The wording for the result value is the same, so would it make sense to have it only once, outside of the numbered list? If yes, the parameter names in (1) and (2) should be aligned to `start, end`. If not, the wording here should use `start1 - end1`.

File: source/elements/oneDPL/source/parallel_api.rst
Comment (akukanov) [inline source/elements/oneDPL/source/parallel_api.rst:711]: I would change the formatting to create a numbered list:
```
1. evaluates the unary predicate :code:`pred` for each position :code:`i` of the sequence
   :code:`[start, end)` and if :code:`pred(start[i]) == true`, it performs the unary operation
   :code:`op(start[i])` and stores the result into :code:`result[i]`. If
```
etc.

File: source/elements/oneDPL/source/parallel_api.rst
Comment (danhoeflinger) [reply source/elements/oneDPL/source/parallel_api.rst]: I agree subtraction is better (but with `end - start`).  fixed.

File: source/elements/oneDPL/source/parallel_api.rst
Comment (danhoeflinger) [reply source/elements/oneDPL/source/parallel_api.rst]: I prefer them separated because at least for now, all the non-high level details of the algorithm are within these numbered list entries.  I think it makes more coherent sense to fully describe the algorithms in one place. 

Also, for the reason you point out that the variables are differently named, I think it is better to separate them. I would prefer to keep the variable naming as is in the prototypes, rather than to make the prototypes less readable to save some space in description.

File: source/elements/oneDPL/source/parallel_api.rst
Comment (danhoeflinger) [reply source/elements/oneDPL/source/parallel_api.rst]: I believe it is fixed now, at least in the github rendering and as I understand the rst syntax.  However, it might be good to get @dcbenito to take a look, I will reach out.

**Review by akukanov — APPROVED** [review]

File: source/elements/oneDPL/source/parallel_api.rst
Comment (dcbenito) [inline source/elements/oneDPL/source/parallel_api.rst:689]: Should e be in quotes? Should the brackets be the same? Right now, it is mixed with squares/parentheses.

File: source/elements/oneDPL/source/parallel_api.rst
Comment (dcbenito) [inline source/elements/oneDPL/source/parallel_api.rst:705]: Why are we using :code:'<value>' in this file and "<value>" in other files? I think that we should standardize this and am partial to "<value>".

Please add a colon after "...the algorithm"

File: source/elements/oneDPL/source/parallel_api.rst
Comment (dcbenito) [inline source/elements/oneDPL/source/parallel_api.rst:708]: Please capitalize evaluates in steps 1 and 2.

**Review by dcbenito — COMMENTED** [review]
Minor changes, but the general text LGTM.

File: source/elements/oneDPL/source/parallel_api.rst
Comment (akukanov) [reply source/elements/oneDPL/source/parallel_api.rst:689]: This `e` thing should be removed, but it's better done in a separate patch.

Comment (dcbenito) [reply source/elements/oneDPL/source/parallel_api.rst:689]: Thank you! I did not know that about the brackets and will keep that in mind for future reviews.

File: source/elements/oneDPL/source/parallel_api.rst
Comment (danhoeflinger) [reply source/elements/oneDPL/source/parallel_api.rst:708]: both are continuing a sentence started prior to the enumerated list, should we still capitalize them?  Or do you think we should reformat the structure to avoid that?

File: source/elements/oneDPL/source/parallel_api.rst
Comment (danhoeflinger) [reply source/elements/oneDPL/source/parallel_api.rst:705]: for :code:` ` vs  ` `, I don't have a good explanation other than staying consistent with the other cases in this file.  If we want to make a formatting change on this topic, I'd suggest we do it for the whole oneDPL spec, and in a separate PR from this.

File: source/elements/oneDPL/source/parallel_api.rst
Comment (dcbenito) [reply source/elements/oneDPL/source/parallel_api.rst:708]: This is a weird one where normal grammar rules don't really apply. Per our and [IDZ's style guide](https://wiki.ith.intel.com/pages/viewpage.action?pageId=1613534372#Formatting&Punctuation-NumberedLists) we format the text using sentence case. We also use a sentence or snippet to introduce the list, so "Depending on the arguments, the algorithm:" is the intro and why we want a colon after "algorithm".

File: source/elements/oneDPL/source/parallel_api.rst
Comment (dcbenito) [reply source/elements/oneDPL/source/parallel_api.rst:705]: After a quick look around, I think that this file is the only one using :code: (for docs anyway). I would prefer that we stay consistent. I think that a new PR for this issue would be good to fix before the next release.

Comment (danhoeflinger) [reply source/elements/oneDPL/source/parallel_api.rst:705]: To be clear, this is the oneDPL specification within the oneAPI specification in UXL, rather than docs which would be tied to an intel release.  
I see a few other instances of `:code:` in `sycl_kernels_api.rst` and `common.rst`, but most are here in this file.  However, I agree it would be good to update with a PR prior to the specification freeze.

File: source/elements/oneDPL/source/parallel_api.rst
Comment (danhoeflinger) [reply source/elements/oneDPL/source/parallel_api.rst]: alright I've applied that pair of changes.

File: source/elements/oneDPL/source/parallel_api.rst
Comment (dcbenito) [reply source/elements/oneDPL/source/parallel_api.rst:705]: That is good to keep in mind!

Would you like me to create a PR? I can have that done by EOD.

Comment (danhoeflinger) [reply source/elements/oneDPL/source/parallel_api.rst:705]: Sure, if you are willing.  Unless there is objection from others for some reasons I don't know about it should be easy to approve.

Comment (dcbenito) [reply source/elements/oneDPL/source/parallel_api.rst:705]: I'll get it started, and if there are objections, we can always delete/modify the PR.

Comment (akukanov) [reply source/elements/oneDPL/source/parallel_api.rst:705]: At the HTML level, this is formatting for backquotes:
```
<code class="docutils literal notranslate">
<span class="pre">using</span> <span class="pre">scalar_type</span> <span class="pre">=</span> <span class="pre">result_type;</span>
</code>
```
and this is the formatting for `::code:`
```
<code class="code docutils literal notranslate">
<span class="pre">namespace</span> <span class="pre">oneapi::dpl::execution</span>
</code>
```
(the examples use different strings - I have not found the same or similar strings for both styles.)

The difference seems to be only in the extra "code" class name in attributes.. I guess it technically allows to set different formatting in CSS or JavaScript, but I do not know if it is used in the oneAPI specification. Visually both styles look the same.

Comment (dcbenito) [reply source/elements/oneDPL/source/parallel_api.rst:705]: I looked through the other products in this repo and we are the only ones using the :code:'value' (single quote)and only in some spots. Most of the instances are "value"(no :code: and with double quotes). We should stick to one. 

I have created a PR, but do not have approval/access to push it. Can someone on this thread give me write access for this repo?

Comment (akukanov) [reply source/elements/oneDPL/source/parallel_api.rst:705]: As far as I can tell, contributions/PRs to this repo are done from forks, not from repo branches.
If you do not want to fork it in your own GitHub account, I can create a PR from my fork.

Comment (dcbenito) [reply source/elements/oneDPL/source/parallel_api.rst:705]: I have followed up via email. Thank you @akukanov

Comment (akukanov) [reply source/elements/oneDPL/source/parallel_api.rst:705]: https://github.com/uxlfoundation/oneAPI-spec/pull/565

File: source/elements/oneDPL/source/parallel_api.rst
Comment (akukanov) [reply source/elements/oneDPL/source/parallel_api.rst:689]: > This e thing should be removed, but it's better done in a separate patch.

Also added to #565.

**Review by timmiesmith — APPROVED** [review]
The description looks good to me.  I'll follow the changes in #565 to continue the discussion.

**Pull request merged**
